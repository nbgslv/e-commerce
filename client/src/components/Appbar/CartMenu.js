import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { makeStyles } from '@material-ui/core/styles';
import MenuItem from '@material-ui/core/MenuItem';
import Divider from '@material-ui/core/Divider';
import MonetizationOnIcon from '@material-ui/icons/MonetizationOn';
import RemoveShoppingCartOutlinedIcon from '@material-ui/icons/RemoveShoppingCartOutlined';
import Menu from '@material-ui/core/Menu';

const useStyles = makeStyles(theme => ({
  root: {
    textDecoration: 'none',
    padding: '0 1% 0 2%',
    color: theme.palette.grey['400'],
    '-o-transition': '.5s',
    '-ms-transition': '.5s',
    '-moz-transition': '.5s',
    '-webkit-transition': '.5s',
    transition: '.5s',
    '&:hover': {
      color: theme.palette.primary.main,
    },
  },
  list: {
    '& > li': {
      padding: '8px 12px 6px 4px',
    },
  },
}));

const CartMenu = ({ anchorEl, open, onClose, emptyCart }) => {
  const classes = useStyles();

  return (
    <Menu
      classes={{ list: classes.list }}
      anchorEl={anchorEl}
      anchorOrigin={{ vertical: 'top', horizontal: 'right' }}
      // id={menuId}
      keepMounted
      transformOrigin={{ vertical: 'top', horizontal: 'right' }}
      open={open}
      onClose={onClose}
    >
      <MenuItem className={classes.root} style={{ display: 'flex', justifyContent: 'center' }}>
        <Link to="/cart" className={classes.root} onClick={onClose}>
          View Cart
        </Link>
      </MenuItem>
      <Divider />
      <MenuItem className={classes.root}>
        <MonetizationOnIcon color="primary" fontSize="small" />
        &nbsp; Purchase
      </MenuItem>
      <MenuItem className={classes.root} onClick={emptyCart}>
        <RemoveShoppingCartOutlinedIcon color="primary" fontSize="small" />
        &nbsp; Empty Cart
      </MenuItem>
    </Menu>
  );
};

CartMenu.propTypes = {
  anchorEl: PropTypes.node.isRequired,
  open: PropTypes.bool.isRequired,
  onClose: PropTypes.func.isRequired,
  emptyCart: PropTypes.func.isRequired,
};

export default CartMenu;
