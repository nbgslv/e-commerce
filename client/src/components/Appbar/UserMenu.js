import { makeStyles } from '@material-ui/core/styles';
import React from 'react';
import MenuItem from '@material-ui/core/MenuItem';
import MeetingRoomOutlinedIcon from '@material-ui/icons/MeetingRoomOutlined';
import Menu from '@material-ui/core/Menu';

const useStyles = makeStyles(theme => ({
  root: {
    textDecoration: 'none',
    padding: '0 1% 0 2%',
    color: theme.palette.grey['400'],
    '-o-transition': '.5s',
    '-ms-transition': '.5s',
    '-moz-transition': '.5s',
    '-webkit-transition': '.5s',
    transition: '.5s',
    '&:hover': {
      color: theme.palette.primary.main,
    },
  },
  list: {
    '& > li': {
      padding: '8px 12px 6px 4px',
    },
  },
}));

const UserMenu = ({ anchorEl, open, onClose, logout }) => {
  const classes = useStyles();

  return (
    <Menu
      classes={{ list: classes.list }}
      anchorEl={anchorEl}
      anchorOrigin={{ vertical: 'top', horizontal: 'right' }}
      // id={menuId}
      keepMounted
      transformOrigin={{ vertical: 'top', horizontal: 'right' }}
      open={open}
      onClose={onClose}
    >
      <MenuItem className={classes.root} onClick={logout}>
        <MeetingRoomOutlinedIcon color="secondary" fontSize="small" />
        &nbsp; Logout
      </MenuItem>
    </Menu>
  );
};

export default UserMenu;
