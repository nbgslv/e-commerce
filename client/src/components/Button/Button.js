import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const ButtonWrapper = styled.button`
  display: flex;
  justify-content: center;
  align-items: center;
  background: transparent;
  color: ${({ color }) => color || 'white'};
  padding: 10px;
  line-height: 2;
  border-radius: 5px;
  font-weight: bold;
  border: 4px solid ${({ color }) => color || 'white'};
  font-size: inherit;
  cursor: pointer;
  text-decoration: none;
`;

const Button = ({ children, color, onClick }) => (
  <ButtonWrapper color={color} onClick={onClick}>
    {children}
  </ButtonWrapper>
);

Button.propTypes = {
  children: PropTypes.arrayOf(PropTypes.object).isRequired,
  color: PropTypes.string,
  onClick: PropTypes.func.isRequired,
};

Button.defaultProps = {
  color: 'white',
};

export default Button;
