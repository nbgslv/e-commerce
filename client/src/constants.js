import gql from 'graphql-tag';

export const GET_CATEGORIES = gql`
  query getCategories {
    categories {
      id
      title
    }
  }
`;

export const COMPLETE_CART = gql`
  mutation completeCart {
    completeCart {
      complete
    }
  }
`;

export const LOGIN_USER = gql`
  mutation loginUser($username: String!, $password: String!) {
    loginUser(userName: $username, password: $password) {
      userName
      token
    }
  }
`;

export const GET_LIMIT = gql`
  query getLimit {
    limit @client
  }
`;

export const GET_PRODUCTS = gql`
  query getProducts($limit: Int) {
    products(limit: $limit) {
      id
      title
      thumbnail
      price
    }
  }
`;

export const GET_CART_TOTAL = gql`
  query getCartTotal {
    cart {
      total
    }
  }
`;

export const GET_CART = gql`
  query getCart {
    cart {
      total
      products {
        id
        title
        thumbnail
      }
    }
  }
`;

export const ADD_TO_CART = gql`
  mutation addToCart($productId: Int!) {
    addToCart(input: { productId: $productId }) {
      total
    }
  }
`;
