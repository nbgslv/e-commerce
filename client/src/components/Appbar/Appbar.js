import React from 'react';
import { useQuery, useMutation, useSubscription } from 'react-apollo';
import Cookies from 'js-cookie';
import { makeStyles, withStyles } from '@material-ui/core/styles';
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import Typography from '@material-ui/core/Typography';
import Button from '@material-ui/core/Button';
import IconButton from '@material-ui/core/IconButton';
import MenuIcon from '@material-ui/icons/Menu';
import Badge from '@material-ui/core/Badge';
import AccountCircle from '@material-ui/icons/AccountCircle';
import ShoppingCart from '@material-ui/icons/ShoppingCart';
import { EMPTY_CART, LOGOUT_USER, GET_USER, CART_CHANGED } from '../../constants';
import { SnackbarContext } from '../../context/snackbarContext';
import { UserContext } from '../../context/UserContext';
import { getUser, emptyCart as emptyLocalCart, getCart, setCart } from '../../utils/localStorage';
import CustomDialog from '../Dialog/CustomDialog';
import CartMenu from './CartMenu';
import UserMenu from './UserMenu';

const useStyles = makeStyles(theme => ({
  root: {
    flexGrow: 1,
  },
  button: {
    padding: '5px',
    margin: '5px',
  },
  menuButton: {
    marginRight: theme.spacing(2),
  },
  title: {
    flexGrow: 1,
  },
}));

const StyledBadge = withStyles(theme => ({
  badge: {
    right: -3,
    top: 13,
    border: `2px solid ${theme.palette.background.paper}`,
    padding: '0 4px',
  },
}))(Badge);

const Appbar = () => {
  const { state, dispatch } = React.useContext(UserContext);
  const { dispatch: snackbarDispatch } = React.useContext(SnackbarContext);
  const { loading, data } = useQuery(GET_USER);
  const { data: updatedCart, loading: cartItemAddedLoading } = useSubscription(CART_CHANGED);
  const [dialogOpen, setDialogOpen] = React.useState(false);

  React.useEffect(() => {
    if (getUser()) {
      if (!cartItemAddedLoading && updatedCart)
        dispatch({
          type: 'UPDATE_CART',
          cart: updatedCart,
        });
      else if (!loading && data) dispatch({ type: 'SET_USER', user: data.getUser });
    } else {
      if (!getCart()) setCart(true);
      dispatch({ type: 'SET_GUEST' });
      Cookies.remove('signedin');
    }
  }, [data, loading, updatedCart, cartItemAddedLoading, state.user.guest]);
  // TODO add loading figure to user appbar right side

  const [anchorElCart, setAnchorElCart] = React.useState(null);
  const [anchorElUser, setAnchorElUser] = React.useState(null);
  const classes = useStyles();

  const handleCartMenuOpen = e => {
    setAnchorElCart(e.currentTarget);
  };

  const handleCartMenuClose = () => {
    setAnchorElCart(null);
  };

  const handleUserMenuOpen = e => {
    setAnchorElUser(e.currentTarget);
  };

  const handleUserMenuClose = () => {
    setAnchorElUser(null);
  };

  const [emptyCart] = useMutation(EMPTY_CART);

  const handleEmptyCart = () => {
    handleCartMenuClose();
    setDialogOpen(true);
  };

  const handleDialogOnClose = async button => {
    setDialogOpen(false);
    if (button === 'right') {
      if (getUser()) await emptyCart();
      else {
        emptyLocalCart();
        dispatch({ type: 'EMPTY_CART' });
      }
      snackbarDispatch({ type: 'SET_EMPTY_CART_SUCCESS_ON' });
    }
  };

  const [logoutUser] = useMutation(LOGOUT_USER);

  const handleLogout = () => {
    handleUserMenuClose();
    const logoutSuccess = logoutUser();
    if (logoutSuccess) {
      Cookies.remove('signedin');
      dispatch({ type: 'REMOVE_USER' });
      snackbarDispatch({ type: 'SET_LOGOUT_SUCCESS_ON' });
    }
  };

  return (
    <div className={classes.root}>
      <AppBar position="fixed">
        <Toolbar>
          <IconButton edge="start" className={classes.menuButton} color="inherit" aria-label="menu">
            <MenuIcon />
          </IconButton>
          <Typography variant="h6" className={classes.title}>
            Photos
          </Typography>
          <div>
            {state.user.guest && (
              <>
                <Button
                  href="/login/"
                  className={classes.button}
                  color="secondary"
                  disableElevation
                >
                  Login
                </Button>
                <Button className={classes.button} variant="outlined" color="secondary">
                  Sign Up
                </Button>
              </>
            )}
            {!state.user.guest && (
              <>
                <IconButton
                  aria-label="account of current user"
                  aria-haspopup="true"
                  onClick={handleUserMenuOpen}
                  color="inherit"
                >
                  <AccountCircle fontSize="large" color="secondary" />
                </IconButton>
                <UserMenu
                  anchorEl={anchorElUser}
                  open={Boolean(anchorElUser)}
                  onClose={handleUserMenuClose}
                  logout={handleLogout}
                />
              </>
            )}
            <IconButton
              edge="end"
              aria-label="shopping cart"
              // aria-controls={menuId}
              aria-haspopup="true"
              onClick={handleCartMenuOpen}
            >
              <StyledBadge badgeContent={state.user.cart.total} color="secondary">
                <ShoppingCart fontSize="large" color="secondary" />
              </StyledBadge>
            </IconButton>
            <CartMenu
              anchorEl={anchorElCart}
              open={Boolean(anchorElCart)}
              onClose={handleCartMenuClose}
              emptyCart={handleEmptyCart}
            />
          </div>
        </Toolbar>
      </AppBar>
      <CustomDialog
        rightButtonLabel="Empty Cart"
        open={dialogOpen}
        onCloseHandler={handleDialogOnClose}
        message="Are you sure you wish to empty the cart?"
        leftButtonLabel="Cancel"
        title="Empty Cart"
      />
    </div>
  );
};

export default Appbar;
