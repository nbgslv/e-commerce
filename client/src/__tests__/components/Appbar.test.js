import '@testing-library/jest-dom';
import ObjectID from 'bson-objectid';
import { shallow, mount } from 'enzyme';
import { render, act } from '@testing-library/react';
import toJson from 'enzyme-to-json';
import React from 'react';
import { MockedProvider } from '@apollo/react-testing';
import { MemoryRouter } from 'react-router-dom';

import Button from '@material-ui/core/Button';
import UserContextProvider, { UserContext } from '../../context/UserContext';
import SnackbarContextProvider from '../../context/snackbarContext';
import Appbar from '../../components/Appbar/Appbar';
import { GET_USER, CART_CHANGED } from '../../constants';

const mocks = [
  {
    request: {
      query: GET_USER,
    },
    result: {
      data: {
        _id: ObjectID,
        firstName: 'John',
        lastName: 'Doe',
        email: 'john@doe.com',
        cart: {
          total: 0,
          products: [],
        },
      },
    },
  },
  {
    request: {
      query: CART_CHANGED,
    },
    result: {
      data: {
        cartChanged: {
          total: 0,
          products: [
            {
              _id: ObjectID,
              title: 'Hot Chips',
              price: 10,
              quantity: 1,
            },
          ],
        },
      },
    },
  },
  {
    request: {
      query: CART_CHANGED,
    },
    result: {
      data: {
        cartChanged: {
          total: 0,
          products: [
            {
              _id: ObjectID,
              title: 'Hot Chips',
              price: 10,
              quantity: 1,
            },
          ],
        },
      },
    },
  },
];

const initializeUserState = {
  user: {
    guest: true,
    cart: {
      total: 0,
      products: [],
    },
  },
};

describe('Appbar', () => {
  let wrapper;
  let userContextMock;
  let mockUserContext;
  beforeEach(() => {
    mockUserContext = {
      dispatch: jest.fn(),
      state: {
        user: {
          guest: true,
          cart: {
            total: 0,
            products: [],
          },
        },
      },
    };
    userContextMock = React.useContext = jest.fn();
    wrapper = mount(
      <MemoryRouter initialEntries={['/']} keyLength={0}>
        <MockedProvider mocks={mocks} addTypename={false}>
          <Appbar />
        </MockedProvider>
      </MemoryRouter>
    );
  });
  afterEach(() => {
    userContextMock.mockReset();
    wrapper.unmount();
  });
  it('shows sign in button', () => {
    console.log(wrapper.debug());
    expect(wrapper.find(Button).filter('[href="/login/"]').at(0).text()).toEqual('Login');
  });
  it('shows sign up button', () => {
    expect(wrapper.find(Button).filter('[name="signupButton"]').at(0).text()).toEqual('Sign Up');
  });
  it('should hide sign up button when user logged in', async () => {
    const promise = () => {
      return new Promise(resolve => {
        setTimeout(() => {
          userContextMock.mockReturnValue({
            ...mockUserContext,
            state: {
              ...mockUserContext.state,
              user: {
                ...mockUserContext.user,
                guest: false,
              },
            },
          });
          resolve(wrapper);
        }, 3000);
      });
    };
    await act(async () => {
      return promise().then(res => {
        expect(res.find(Button).filter('[name="signupButton"]').at(0).exists()).toBe(false);
      });
    });
  });
});
