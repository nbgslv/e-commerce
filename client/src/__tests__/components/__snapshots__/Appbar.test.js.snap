// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Appbar renders without crashing 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "canGo": [Function],
      "createHref": [Function],
      "entries": Array [
        Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
      ],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "index": 0,
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <MockedProvider
    addTypename={false}
    mocks={
      Array [
        Object {
          "request": Object {
            "query": Object {
              "definitions": Array [
                Object {
                  "directives": Array [],
                  "kind": "OperationDefinition",
                  "name": Object {
                    "kind": "Name",
                    "value": "getUser",
                  },
                  "operation": "query",
                  "selectionSet": Object {
                    "kind": "SelectionSet",
                    "selections": Array [
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "getUser",
                        },
                        "selectionSet": Object {
                          "kind": "SelectionSet",
                          "selections": Array [
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "_id",
                              },
                              "selectionSet": undefined,
                            },
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "firstName",
                              },
                              "selectionSet": undefined,
                            },
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "lastName",
                              },
                              "selectionSet": undefined,
                            },
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "cart",
                              },
                              "selectionSet": Object {
                                "kind": "SelectionSet",
                                "selections": Array [
                                  Object {
                                    "alias": undefined,
                                    "arguments": Array [],
                                    "directives": Array [],
                                    "kind": "Field",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "total",
                                    },
                                    "selectionSet": undefined,
                                  },
                                  Object {
                                    "alias": undefined,
                                    "arguments": Array [],
                                    "directives": Array [],
                                    "kind": "Field",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "products",
                                    },
                                    "selectionSet": Object {
                                      "kind": "SelectionSet",
                                      "selections": Array [
                                        Object {
                                          "alias": undefined,
                                          "arguments": Array [],
                                          "directives": Array [],
                                          "kind": "Field",
                                          "name": Object {
                                            "kind": "Name",
                                            "value": "_id",
                                          },
                                          "selectionSet": undefined,
                                        },
                                        Object {
                                          "alias": undefined,
                                          "arguments": Array [],
                                          "directives": Array [],
                                          "kind": "Field",
                                          "name": Object {
                                            "kind": "Name",
                                            "value": "title",
                                          },
                                          "selectionSet": undefined,
                                        },
                                        Object {
                                          "alias": undefined,
                                          "arguments": Array [],
                                          "directives": Array [],
                                          "kind": "Field",
                                          "name": Object {
                                            "kind": "Name",
                                            "value": "price",
                                          },
                                          "selectionSet": undefined,
                                        },
                                        Object {
                                          "alias": undefined,
                                          "arguments": Array [],
                                          "directives": Array [],
                                          "kind": "Field",
                                          "name": Object {
                                            "kind": "Name",
                                            "value": "quantity",
                                          },
                                          "selectionSet": undefined,
                                        },
                                      ],
                                    },
                                  },
                                ],
                              },
                            },
                          ],
                        },
                      },
                    ],
                  },
                  "variableDefinitions": Array [],
                },
              ],
              "kind": "Document",
              "loc": Object {
                "end": 213,
                "start": 0,
              },
            },
          },
          "result": Object {
            "data": Object {
              "_id": [Function],
              "cart": Object {
                "products": Array [],
                "total": 0,
              },
              "email": "john@doe.com",
              "firstName": "John",
              "lastName": "Doe",
            },
          },
        },
        Object {
          "request": Object {
            "query": Object {
              "definitions": Array [
                Object {
                  "directives": Array [],
                  "kind": "OperationDefinition",
                  "name": Object {
                    "kind": "Name",
                    "value": "cartChanged",
                  },
                  "operation": "subscription",
                  "selectionSet": Object {
                    "kind": "SelectionSet",
                    "selections": Array [
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "cartChanged",
                        },
                        "selectionSet": Object {
                          "kind": "SelectionSet",
                          "selections": Array [
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "total",
                              },
                              "selectionSet": undefined,
                            },
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "products",
                              },
                              "selectionSet": Object {
                                "kind": "SelectionSet",
                                "selections": Array [
                                  Object {
                                    "alias": undefined,
                                    "arguments": Array [],
                                    "directives": Array [],
                                    "kind": "Field",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "_id",
                                    },
                                    "selectionSet": undefined,
                                  },
                                  Object {
                                    "alias": undefined,
                                    "arguments": Array [],
                                    "directives": Array [],
                                    "kind": "Field",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "title",
                                    },
                                    "selectionSet": undefined,
                                  },
                                  Object {
                                    "alias": undefined,
                                    "arguments": Array [],
                                    "directives": Array [],
                                    "kind": "Field",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "price",
                                    },
                                    "selectionSet": undefined,
                                  },
                                  Object {
                                    "alias": undefined,
                                    "arguments": Array [],
                                    "directives": Array [],
                                    "kind": "Field",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "quantity",
                                    },
                                    "selectionSet": undefined,
                                  },
                                ],
                              },
                            },
                          ],
                        },
                      },
                    ],
                  },
                  "variableDefinitions": Array [],
                },
              ],
              "kind": "Document",
              "loc": Object {
                "end": 152,
                "start": 0,
              },
            },
          },
          "result": Object {
            "data": Object {
              "cartChanged": Object {
                "products": Array [
                  Object {
                    "_id": [Function],
                    "price": 10,
                    "quantity": 1,
                    "title": "Hot Chips",
                  },
                ],
                "total": 0,
              },
            },
          },
        },
        Object {
          "request": Object {
            "query": Object {
              "definitions": Array [
                Object {
                  "directives": Array [],
                  "kind": "OperationDefinition",
                  "name": Object {
                    "kind": "Name",
                    "value": "cartChanged",
                  },
                  "operation": "subscription",
                  "selectionSet": Object {
                    "kind": "SelectionSet",
                    "selections": Array [
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "cartChanged",
                        },
                        "selectionSet": Object {
                          "kind": "SelectionSet",
                          "selections": Array [
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "total",
                              },
                              "selectionSet": undefined,
                            },
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "products",
                              },
                              "selectionSet": Object {
                                "kind": "SelectionSet",
                                "selections": Array [
                                  Object {
                                    "alias": undefined,
                                    "arguments": Array [],
                                    "directives": Array [],
                                    "kind": "Field",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "_id",
                                    },
                                    "selectionSet": undefined,
                                  },
                                  Object {
                                    "alias": undefined,
                                    "arguments": Array [],
                                    "directives": Array [],
                                    "kind": "Field",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "title",
                                    },
                                    "selectionSet": undefined,
                                  },
                                  Object {
                                    "alias": undefined,
                                    "arguments": Array [],
                                    "directives": Array [],
                                    "kind": "Field",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "price",
                                    },
                                    "selectionSet": undefined,
                                  },
                                  Object {
                                    "alias": undefined,
                                    "arguments": Array [],
                                    "directives": Array [],
                                    "kind": "Field",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "quantity",
                                    },
                                    "selectionSet": undefined,
                                  },
                                ],
                              },
                            },
                          ],
                        },
                      },
                    ],
                  },
                  "variableDefinitions": Array [],
                },
              ],
              "kind": "Document",
              "loc": Object {
                "end": 152,
                "start": 0,
              },
            },
          },
          "result": Object {
            "data": Object {
              "cartChanged": Object {
                "products": Array [
                  Object {
                    "_id": [Function],
                    "price": 10,
                    "quantity": 1,
                    "title": "Hot Chips",
                  },
                ],
                "total": 0,
              },
            },
          },
        },
      ]
    }
  >
    <Appbar />
  </MockedProvider>
</Router>
`;
